[jayk.yang] 지하철 노선도 step3 리뷰 요청 드립니다

tttuer opened the pull request:
지금 도메인으로 뺄 수 있을만한 부분은 분리해봤습니다만 더 분리할 수 있는 부분이 있다면 피드백 주시면 감사하겠습니다!
또한 dao와 서비스를 분리해봤는데 잘 분리됐는지 피드백 부탁드립니다. 감사합니다!

Ye-Ji-Kim committed:
feat: CRUD 구현

Station CRUD, Line CRUD 구현

Ye-Ji-Kim committed:
feat: LineResponse default 생성자 추가

Ye-Ji-Kim committed:
feat: step2 지하철 구간 추가 API구현

tttuer committed:
feat: 1단계 실습

tttuer committed:
Merge branch 'step1' of https://github.com/Ye-Ji-Kim/spring-subway-admin-kakao into step1

Ye-Ji-Kim committed:
feat: 구간 등록 API 에러 수정

tttuer committed:
Merge branch 'step1' of https://github.com/Ye-Ji-Kim/spring-subway-admin-kakao into step1

Ye-Ji-Kim committed:
feat: step2 지하철 구간 추가 API 최종 수정

Ye-Ji-Kim committed:
feat: step3 jdbc로 변경

tttuer committed:
Merge branch 'step1' of https://github.com/Ye-Ji-Kim/spring-subway-admin-kakao into step1

tttuer committed:
feat: 버그 수정

Test실행 했을때 오류 수정

Ye-Ji-Kim committed:
feat: step3 구간 추가 API 에러 수정 및 구간 삭제 API 구현

tttuer committed:
feat: 안쓰는 것들 제거

tttuer committed:
Merge branch 'step1' of https://github.com/Ye-Ji-Kim/spring-subway-admin-kakao into step1

tttuer committed:
feat: 버그 수정

SectionService에서 삭제 시 오류 나던 부분 수정

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,95 @@
+        if (newLine == null) {
+            return ResponseEntity.badRequest().build();
+        }
+        sectionService.insert(new Section(newLine.getId(), lineRequest.getUpStationId(), lineRequest.getDownStationId(), lineRequest.getDistance()));
```
위 코드로직은 하나의 트랜젝션으로 처리되어야 하는 로직입니다. 
Line 저장로직 ( Section 을 포함 ) 을 담당해줄 수 있는 객체로 이동하는게 좋겠습니다.
현재 코드구성 상으로는 Service 쪽으로 이동시키는게 좋겠습니다.

sehan commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,73 @@
+    public Line save(Line line) {
+        List<Line> lines = findLineByName(line.getName());
+        if (lines.size() > 0) {
+            return null;
```
1. null 을 return 하는것 보다 custom exception 을 던져서 저장이 되지 않은 이유를 명확히 해주는것이 좋겠습니다.
null 은 불가피할 때가 아니면 return 값으로 사용하지 않는것이 좋습니다. caller 측에서 null 체크를 상시 체크해야 하는 부담이 발생하고 NPE 발생 위험이 높아 집니다.

2. "line 이름은 중복을 허용하지 않는다" 라는 기능명세는 도메인 측에서 로직을 구현하는것이 어떨까요?  persistent 와 도메인을 분리한다고 할때 위 로직이 persistent 에 있으면 persistent 마다 로직을 추가해야 합니다. 도메인측에 로직을 구현하고 persistent 는 중복체크를 위한 메소드를 제공해주면 어떨까요?

sehan commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,73 @@
+
+        return new Line(keyHolder.getKey().longValue(), line.getName(), line.getColor());
+    }
+
```
JdbcTemplate 을 사용한 dao 매소드들이 간결하고 깔끔합니다. 👍 👍 👍

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,95 @@
+    @GetMapping("/lines/{lineId}")
+    public ResponseEntity<LineResponse> showLine(@PathVariable Long lineId) {
+        Line line = lineDao.findLineById(lineId);
+        if (line == null) {
```
null 을 피할수 없는 경우라면 Optional 사용을 한번 고려 해보는것도 좋습니다.
null 의미가 모호하기 때문에 custom exception 으로 처리해보는것도 같이 한번 고려해보세요 😊

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,95 @@
+
+    @GetMapping(value = "/lines", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<List<LineResponse>> showLines() {
+        List<LineResponse> responses = lineDao.findAll().stream()
```
Response DTO 로 객체 변환하는 코드들이 많이 등장하고 있습니다.  이 객체를 convert 해주는 역할을 유틸성으로 분리해 주면 재사용성이 높아지고 가독성도 개선됩니다. 객체 멤버들에 변경이 생겨도 convert 역할을 담당하는 객체로 변경범위가 줄어 들어 확장성에도 도움이 됩니다.

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,95 @@
+
+    @PostMapping("/lines/{lineId}/sections")
+    public ResponseEntity<LineResponse> createSection(@PathVariable Long lineId, @RequestBody SectionRequest sectionRequest) {
+        Section newSection = new Section(lineId, sectionRequest.getUpStationId(), sectionRequest.getDownStationId(), sectionRequest.getDistance());
```
SectionRequest 로 Section 을 생성하기 위해 많은 parameter 들을 주고 받고 있습니다.  parameter 변경이 발생할때 마다 Section 객체를 생성하는 코드들을 수정해야 합니다. Factory 를 두어 변경범위를 한곳으로 집중시키면 코드변경시에 변경범위가 줄어들어 이점이 있습니다.

tttuer replied:
Section객체 안에 Request를 받아서 생성자에서 처리를 해주는 식으로 처리해봤습니다. 혹시 말씀해주신 의도와 다르다면 다시 한번 피드백 해주시면 감사하겠습니다!

sehan replied:
SectionRequest 를 Section 생성자 인자로 받게 되면 Section 이 SectionRequest 에 의존하게 됩니다.
SectionRequest 는 UI 에 속하는 객체이고 Section 는 도메인에 속하기 때문에 도메인 -> UI 로 의존을 갖지 않도록 하는게 좋습니다.
위에서 의도했던것은 다음과 같은 방식이었습니다. ☺️
```
Section section = SectionFactory.create(SectionRequest)  
```

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,95 @@
+        try {
+            sectionService.insert(newSection);
+        } catch (IllegalArgumentException illegalArgumentException) {
+            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
```
IllegalArgumentException 이 원인이라면 사용자의 잘못된 입력으로 인한 오류로 판단되니 500 에러보다는 400 BadRequest 가 더 적합해 보입니다.

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,95 @@
+    public ResponseEntity<LineResponse> deleteSection(@PathVariable Long lineId, @RequestParam Long stationId) {
+        try {
+            sectionService.delete(lineId, stationId);
+        } catch (IllegalArgumentException illegalArgumentException) {
```
exception 변수는 긴 이름보다는 e, 또는 ex 정도로 표기해도 충분합니다.

```suggestion
        } catch (IllegalArgumentException e) {
```

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,95 @@
+                .map(station -> new StationResponse(station.getId(), station.getName()))
+                .collect(Collectors.toList());
+
+        LineResponse lineResponse = new LineResponse(line.getId(), line.getName(), line.getColor(), stationResponses);
```
LineResponse 가 상황에 따라 stations 을 포함하기도 하고 그렇지 않기도 합니다.
Data 가 상황에 따라 달라지게 되면 클라이언트 측에서 stations 이 포함되어 있는지 여부를 혼돈할 여지가 있고 stations 존재 여부를 늘 체크하며 사용해야 합니다. 

용도에 맞게 Response DTO 를 각각 만들어서 사용하는것이 좋습니다.

sehan commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,95 @@
+
+    @DeleteMapping("/lines/{id}")
+    public ResponseEntity deleteStation(@PathVariable Long id) {
+        lineDao.deleteById(id);
```
line 삭제시 line 에 포함되는 section 들도 삭제가 되어야 합니다. 
두 작업은 하나의 트랜젝션으로 처리되어야 합니다.

sehan commented on src/main/java/subway/line/SectionService.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    public void insert(Section section) {
+        section.checkValidInsert(showAll(section.getLineId()));
```
추가할 Section 에게 기존 Sections 를 건네주고 추가가능여부를 확인 할 수도 있지만
기존 Sections 에게 새로운 Section 을 추가할 수 있는지 확인 해 볼 수 도 있을거 같습니다.
또는 기존 Sections 의 상태를 Line 이 관리하고 있다면 Line 에게 도움을 요청하는 방식으로 풀어 볼 수 도 있습니다.
로직을 도메인객체로 이동시킬 방법은 없는지 한번 고민해주세요 ☺️

sehan commented on src/main/java/subway/line/SectionService.java:
```diff
@@ -0,0 +1,120 @@
+        sectionDao.update(new Section(downStationMatchSection.getId(), section.getLineId(), downStationMatchSection.getUpStationId(), section.getUpStationId(), newDistance));
+    }
+
+    public int delete(Long lineId, Long stationId) {
```
삭제 로직은 하나의 트랜젝션으로 동작해야 합니다. 그래야 실행중 문제가 발생하더라도 데이터 일관성이 유지됩니다.

tttuer replied:
이부분에서 하나의 트랜젝션으로 동작해야한다는 이야기가 잘 와닿지 않습니다. 경우에 따라서 다르게 삭제를 하는것을 지양하라는 말씀이신가요?

sehan replied:
LineService.deleteById() 에 작성하려고 했던 거였는데 다른곳에 작성을 해놨네요 😅
Line 삭제시 관련 section 들과 함께 삭제되도록 하나의 트랜젝션으로 묶어서 처리하자는 코멘트 였습니다.

sehan reviewed (changes requested):
안녕하세요. 지하철 미션을 함께 하게된 니오입니다. ☺️

요구사항 따라 기능구현을 잘 해주셨습니다. 👍 👍 👍 
몇가지 부분에서 아쉬운 점들이 보여 코멘트 남겼습니다. 
현재 Service 객체에 거의 모든 로직들 구현되어있습니다. 로직들을 도메인쪽으로 이동시켜 Service 를 얇게 하고 도메인을 두껍게 가져가 보는 작업을 진행해 보시기 바랍니다.
현재 도메인 객체들에 일부 로직이 있지만  table 에 1:1 로 매핑되어 있는 상태로 실제로는 entity 입니다. 다시 말해서 entity 만 있을뿐 도메인객체는 없는것과 같습니다. 도메인 객체들이 더 많은 역할을 할 수 있도록 더 개선해 볼것을 권합니다. 

그외 몇가지 내용들은 코멘트에 남겨 두었습니다. 확인 부탁드려요 

수정요청 드리겠습니다.

tttuer committed:
refactor: 피드백 반영

도메인을 생성하여 서비스 부분을 축소, 라인 생성시 section도 같이 생성하도록 service에 트랜잭션 추가, exception 분류 및 null 리턴하지 않도록 수정, null을 확인하는 대신 optional을 이용하여 익셉션 발생하도록 수정, Response를 생성해주는 ResponseUtil 추가, 익셉션 변수 이름 수정, line 삭제시 그 안의 섹션들도 같이 삭제되도록 트랜잭션 추가

tttuer committed:
refactor: 피드백 반영

SectionRequest를 Section이 받아서 생성자에서 처리하도록 수정

sehan commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -27,10 +27,7 @@ public LineDao(JdbcTemplate jdbcTemplate) {
-        if (lines.size() > 0) {
-            return null;
-        }
+        Lines lines = new Lines(findLineByName(line.getName()));
```
Line 을 생성하기 위해 이름중복을 체크하는 로직은 Dao 가 아닌 도메인쪽에서 처리하는 것이 좋습니다.
Dao 에 로직을 추가할 수도 있겠지만 비지니스 로직은 최대한 도메인 쪽으로 이동 시켜주세요

추가. 이름중복검증을 위해 Lines 객체를 만들 필요까지는 없어 보입니다.

sehan commented on src/main/java/subway/line/SectionDao.java:
```diff
@@ -64,4 +64,9 @@ public int delete(Section section) {
         return jdbcTemplate.update(sql, section.getId(), section.getUpStationId(), section.getDownStationId());
     }
 
+    public int delete(Long lineId) {
```
delete(long id) 또는 deleteById(long id) 같은 이름은 section id 로 section 을 삭제하는 것으로 혼도될 수 있습니다.
deleteByLineId(long lineId) 와 같이 명확한 이름이 좋습니다.

sehan reviewed (approved):
서비스에 있던 로직들이 조금씩 분리되어 나가니 한결 좋아졌습니다. 👍 

하지만 4단계 layered architecture 를 고려하여 리팩토링을 진행하기 위해서는 아직도 갈길이 많이 남아있습니다. 앞선 자동차/로또 미션에서 경험했던것과 같이 이번 미션에서도 도메인객체들은 다른 layer 에 의존없이 단위테스트 작성이 가능해야 합니다. 이 부분을 고려해서 4단계 잘 진행하시기 바랍니다.

이번 step 은 merge 하도록 하겠습니다.  고생하셨습니다.
